8331194: NPE in ArrayCreationTree.java with -XX:-UseCompressedOops

JohnTortugo opened the pull request:
Please, review this PR to fix issue in debug information serialization encountered when RAM reduces a Phi for which one of the inputs is an object already scalar replaced.

**Details:**

Consider class `Picture` that has two reference fields, `first` and `second` of type `Point`. In a random method in the application an object `obj` of this class is created, and the fields of this object are initialized such that `first` is assigned a new object whereas `second` receives the output of a `Phi` node merging the object assigned to `first` and some other allocation. Also, assumes `obj` is used as debug information in an `uncommon_trap` and none of these objects escapes. I.e., we have a scenario like this:

```
Picture obj       = new Picture(); // allocation A
        obj.first = new Point(); // allocation B

Point    p2    = obj.first;
if (<cond>) p2 = new Point(); // allocation C

obj.second = p2;

<trap>
```

After one iteration of EA+SR, Allocation `A` will be scalar replaced and debug information in `<Trap>` adjusted accordingly. The description of field `second` in the debug information on `<Trap>` will, however, still involve a `Phi` node between allocation `B` and `C`. In the next iteration of EA+SR the `Phi` node for field `second` will be reduced by RAM and debug information will be adjusted accordingly. So far nothing is wrong.

The issue happens because the existing code in `Process_OopMap_Node` to serialize debug information was missing a check. Simply, the check for setting `is_root` of an `ObjectValue` wasn't checking that the `ObjectValue` might be a description of a field of a scalar replaced object. It was only checking whether `ObjectValue` was a local, stack or monitor. Consequently, the allocation assigned to `obj.first` (yes, `first`) was **not** being marked as _root_.

But the issue only manifested if the `<trap>` was exercised **AND** the result of `<cond>` was _true_. If the result of `<cond>` was `false` when the trap was exercised, then no problem would happen. The reason is, when `<cond>` is _true_ the `select` method in [ObjectMergeValue](https://github.com/openjdk/jdk/blob/1472124489c841642996ae984e21c533ffec8091/src/hotspot/share/code/debugInfo.cpp#L237) would flag, correctly, that the allocation inside the `if` needs to be rematerialized and the other input of the ObjectMergeValue shouldn't be rematerialized because `_is_root == false`, meaning it's just a candidate for rematerialization.

Fixing the check for setting `_is_root` solved the problem. The `-XX:-UseCompressedOops` wasn't directly related to the problem, it just caused `DecodeN`, `EncodeP` nodes to not show up in the graph and RAM consider the Phi for reduction.


<!--
Replace this text with a description of your pull request (also remove the surrounding HTML comment markers).
If in doubt, feel free to delete everything in this edit box first, the bot will restore the progress section as needed.
-->

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8331194](https://bugs.openjdk.org/browse/JDK-8331194): NPE in ArrayCreationTree.java with -XX:-UseCompressedOops (**Bug** - P2)


### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20087/head:pull/20087` \
`$ git checkout pull/20087`

Update a local copy of the PR: \
`$ git checkout pull/20087` \
`$ git pull https://git.openjdk.org/jdk.git pull/20087/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20087`

View PR using the GUI difftool: \
`$ git pr show -t 20087`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20087.diff">https://git.openjdk.org/jdk/pull/20087.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20087#issuecomment-2218804204)

eme64 commented on src/hotspot/share/opto/machnode.hpp:
```diff
@@ -867,6 +867,10 @@ class MachSafePointNode : public MachReturnNode {
     assert(verify_jvms(jvms), "jvms must match");
     return in(_jvmadj + jvms->monitor_box_offset(idx));
   }
+  Node *scalarized(const JVMState* jvms, uint idx) const {
```
```suggestion
  Node* scalarized(const JVMState* jvms, uint idx) const {
```

eme64 commented on src/hotspot/share/opto/output.cpp:
```diff
@@ -974,6 +974,24 @@ bool PhaseOutput::contains_as_owner(GrowableArray<MonitorValue*> *monarray, Obje
+// Determine if there is a scalar replaced object description represented by 'ov'.
+bool PhaseOutput::contains_as_scalarized(JVMState* jvms, MachSafePointNode* sfn,
+                                         GrowableArray<ScopeValue*>* objs,
+                                         ObjectValue *ov) const {
```
```suggestion
                                         ObjectValue* ov) const {
```

eme64 commented on test/hotspot/jtreg/compiler/c2/TestReduceAllocationAndNestedScalarized.java:
```diff
@@ -0,0 +1,79 @@
+ *          of the Phi is not the _current_ output of the Phi but said input
+ *          needs to be rematerialized because it's used regardless of the
+ *          Phi output.
+ * @requires vm.bits == 64 & vm.flagless & vm.compiler2.enabled
+ * @run main/othervm -XX:CompileCommand=dontinline,*TestReduceAllocationAndNestedScalarized*::test
+ *                   -XX:CompileCommand=compileonly,*TestReduceAllocationAndNestedScalarized*::test
+ *                   -XX:CompileCommand=compileonly,*Picture*::*init*
+ *                   -XX:CompileCommand=compileonly,*Point*::*init*
+ *                   -XX:CompileCommand=exclude,*Unloaded*::*
+ *                   -XX:-TieredCompilation
+ *                   -XX:-UseCompressedOops
+ *                   -Xcomp
+ *                   -server
+ *                   compiler.c2.TestReduceAllocationAndNestedScalarized
```
Drive by comment: are all these `requires` absolutely necessary? Could there be a run without them? Is the test expected to fail with graal or C1? Or any external flags?

Could we have a run without some of the flags you list here? Just so that we have wider coverage, especially in combination with other flags that can be passed from the outside.

JohnTortugo replied:
This test is intended to "test" a transformation in C2, it won't fail if run in C1, but it won't actually test what's intended for. The `flagless` option is to make sure that C2 produce an IR graph that will trigger the optimization, same reasoning for all the "CompileCommand" options. I may remove the "vm.bits" part, I included it because of the "-UseCompressedOops".

eme64 commented:
@JohnTortugo 

> This test is intended to "test" a transformation in C2, it won't fail if run in C1, but it won't actually test what's intended for. The flagless option is to make sure that C2 produce an IR graph that will trigger the optimization, same reasoning for all the "CompileCommand" options. I may remove the "vm.bits" part, I included it because of the "-UseCompressedOops".

The `flagless` requires will just mean that your test is not run if there are flags. So we still rely on the test configuration having some run without additional flags. So your requires seems pointless, especially because your test does not crash without it. Unless your test takes a lot of time and you are trying to conserve runtime.

About flags: I would just have a second run without any flags.

vnkozlov commented:
You can use `-XX:+IgnoreUnrecognizedVMOptions` flag instead of require "vm.bits".
Yes, do second run without flags.

I also suggest put test into corresponding directory `compiler/escapeAnalysis`. `compiler/c2` is used for legacy or some C2 general tests.

vnkozlov commented on src/hotspot/share/opto/machnode.hpp:
```diff
@@ -867,6 +867,10 @@ class MachSafePointNode : public MachReturnNode {
     assert(verify_jvms(jvms), "jvms must match");
     return in(_jvmadj + jvms->monitor_box_offset(idx));
   }
+  Node* scalarized(const JVMState* jvms, uint idx) const {
```
May be rename to `scalarized_obj()`.

vnkozlov commented on src/hotspot/share/opto/output.cpp:
```diff
@@ -974,6 +974,24 @@ bool PhaseOutput::contains_as_owner(GrowableArray<MonitorValue*> *monarray, Obje
+                                         ObjectValue* ov) const {
+  for (int i = 0; i < jvms->scl_size(); i++) {
+    Node* n = sfn->scalarized(jvms, i);
+    if (!n->is_SafePointScalarObject()) {
```
Needs comment explaining what other Node type could be here. May be assert check for them in `else` case.

vnkozlov reviewed (commented):
The fix looks fine. I have few comments.

JohnTortugo commented:
@vnkozlov @eme64 - I moved all tests related to "RAM" to the suggested folder and modified the test, added in this PR, to remove the flags as you suggested.

Please, let me ask more about configuring the test to run without flags. Without the flags, for instance, the CompileCommands, the IR graph will very likely not be in the shape that trigger the problem. Even with the CompileCommands, if some other flags included in the list of flags to run the test, the IR graph may also not be in the shape that will trigger the problem. Why run the test if it doesn't trigger the problem it was intended for? I'm probably missing something here.

vnkozlov reviewed (changes requested):
Please, reverse movement of old tests. We need to backport this fix into JDK 23 and for that we need only related test.
You can move them in separate RFE in JDK 24 (current mainline).

vnkozlov commented:
> Why run the test if it doesn't trigger the problem it was intended for? I'm probably missing something here.

We hope to catch other issues with each new test. We are running with variety of flags combinations set by testing environment and a test may fail for different reason. It is all about code shape variations and flags combinations.

JohnTortugo commented:
> Please, reverse movement of old tests. We need to backport this fix into JDK 23 and for that we need only related test. You can move them in separate RFE in JDK 24 (current mainline).

I reverted the move. I had forgot about the backport.

vnkozlov reviewed (approved):
Good.

eme64 commented on test/hotspot/jtreg/compiler/escapeAnalysis/TestReduceAllocationAndNestedScalarized.java:
```diff
@@ -0,0 +1,81 @@
+ *          of the Phi is not the _current_ output of the Phi but said input
+ *          needs to be rematerialized because it's used regardless of the
+ *          Phi output.
+ * @requires vm.compiler2.enabled
```
Again, I think you do not neet do restrict test execution to C2. The `requires` does not make C2 appear magically, but rather it just means the test is not executed if C2 is not available. So why restrict the test? Maybe some other compiler also has a similar or totally different bug triggered with your test. Would that not be nice?

eme64 replied:
Now that you have `-XX:+IgnoreUnrecognizedVMOptions`, you do not have to be worried that some other compiler crashes because of some unrecognized flag.

JohnTortugo replied:
> Would that not be nice?

Fingers crossed!

eme64 commented on test/hotspot/jtreg/compiler/escapeAnalysis/TestReduceAllocationAndNestedScalarized.java:
```diff
@@ -0,0 +1,81 @@
+ * @run main/othervm compiler.c2.TestReduceAllocationAndNestedScalarized
+ */
+
+package compiler.c2;
```
You moved your test file. The package name should either be removed or adjusted accordingly ;)

vnkozlov reviewed (approved):
Update is good.

vnkozlov commented:
I will run testing with latest version. Tobias ran with version 0 and all passed. Since then there were only cosmetic changes but we still need to verify it.

JohnTortugo commented:
Sounds good. Thank you!

vnkozlov commented:
My testing passed.

@JohnTortugo, can you check why only 5 tests ran in GHA testing?

JohnTortugo commented:
@vnkozlov - I'm looking into that and I'll update here.

JohnTortugo commented:
/integrate

vnkozlov commented:
@eme64  I you fine with current version?

eme64 commented on test/hotspot/jtreg/compiler/escapeAnalysis/TestReduceAllocationAndNestedScalarized.java:
```diff
@@ -0,0 +1,80 @@
+ *                   -Xcomp
+ *                   -server
+ *                   compiler.escapeAnalysis.TestReduceAllocationAndNestedScalarized
+ * @run main/othervm compiler.escapeAnalysis.TestReduceAllocationAndNestedScalarized
```
```suggestion
 * @run main compiler.escapeAnalysis.TestReduceAllocationAndNestedScalarized
```
You can run it directly in the main vm, with the external flags applied. othervm creates an extra vm which is unnecessary if you do not have additional arguments. So your first run needs othervm, but not your second ;)

eme64 commented on src/hotspot/share/opto/output.cpp:
```diff
@@ -974,6 +974,27 @@ bool PhaseOutput::contains_as_owner(GrowableArray<MonitorValue*> *monarray, Obje
+      continue;
+    }
+
+    ObjectValue* other = (ObjectValue*) sv_for_node_id(objs, n->_idx);
```
Is the cast here necessary? I see them generally in the file... but not sure why.
```
ObjectValue*
PhaseOutput::sv_for_node_id(GrowableArray<ScopeValue*> *objs, int id) {
```

vnkozlov replied:
Because `objs` array may contains 3 types of objects: `ScopeValue`, `ObjectValue` and `ObjectMergeValue`.
I would leave this code as it is for this PR but suggest to file followup RFE to clean this up.
Instead of such casts we should use `as_ObjectValue()` and `as_ObjectMergeValue()` which have asserts to check type.

JohnTortugo replied:
Thank you @eme64 , @vnkozlov for reviewing. I created this RFE to remove the unnecessary casts: https://bugs.openjdk.org/browse/JDK-8336495

eme64 commented:
@vnkozlov @JohnTortugo I'm not familiar enough with this part of the code, just gave some code style and testing suggestions. I was given something else that has higher priority, so maybe some one else can give the VM changes a review?

vnkozlov reviewed (approved):
Update is good.

JohnTortugo commented:
/integrate

vnkozlov commented:
/sponsor

TobiHartmann commented:
/backport :jdk23
